android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```diff
@@ -26,51 +28,145 @@
+    public Single<UploadValidationStatus> uploadToS3(String filename) {
+        UploadFile uploadFile = uploadDAO.getUploadFile(filename);
+        if (uploadFile == null) {
+            LOG.warn("Could not find uploadFile for filename: " + filename);
+
+            return Single.just(null);
+        }
 
-        return Single.zip(uploadFileSingle, uploadSessionSingle, this::uploadToS3).flatMap(i -> i);
+        return getUploadSession(uploadFile)
+                .flatMap(uploadSession -> getUploadValidationStatus(uploadSession.getId()))
+                .flatMap((uploadValidationStatus) -> {
+                    if (uploadValidationStatus == null) {
+                        LOG.info("Could not find uploadValidationStatus for filename: " + filename);
+                        return Single.just(null);
+                    }
+
+                    if (UploadStatus.SUCCEEDED == uploadValidationStatus.getStatus()) {
+                        LOG.info("Cleaning up successful upload with filename: " + filename
+                                + ", id: " + uploadValidationStatus.getId());
+                        uploadDAO.removeUploadAndSession(filename);
+                        return Single.just(null);
+                    } else if (UploadStatus.VALIDATION_FAILED == uploadValidationStatus.getStatus()) {
+                        LOG.info("Cleaning up validation-failed upload with filename: " + filename
+                                + ", id: " + uploadValidationStatus.getId());
+                        uploadDAO.removeUploadAndSession(filename);
+                        return Single.just(null);
+                    }
+
+                    return getUploadSession(uploadFile)
+                            .flatMapCompletable(uploadSession -> uploadToS3(uploadFile, uploadSession))
+                            .andThen(getUploadValidationStatus((uploadValidationStatus.getId())));
+                }).doOnSuccess(uploadValidationStatus -> {
+                    LOG.debug("uploadToS3 succeded, uploadValidationStatus: " + uploadValidationStatus);
+                    if (uploadValidationStatus == null) {
+                        LOG.warn("null uploadValidationStatus");
+                    }
+
+                }).onErrorReturn(throwable -> {
+                    LOG.warn("Failed attempt to upload pending file: " + filename, throwable);
+                    if (isPermanentFailure(uploadFile, throwable)) {
+                        uploadDAO.removeUploadAndSession(filename);
+                    }
+                    return null;
+                });
     }
 
-    @NonNull
-    Single<UploadValidationStatus> uploadToS3(UploadFile uploadFile, UploadSession session) {
+    // determines whether an upload failure is permanent, i.e. no point in retrying
+    boolean isPermanentFailure(UploadFile uploadFile, Throwable throwable) {
+        if (throwable instanceof IllegalStateException) {
+            // some pre-condition failed
+            return true;
```

Comment:
Can you elaborate more in comments about specifically what would cause an IllegalStateException here?